WalletWasabi/WabiSabi/Backend/Banning/CoinVerifier.cs

```diff
@@ -28,101 +34,111 @@ internal CoinVerifier(CoinJoinIdStore coinJoinIdStore, CoinVerifierApiClient api
-		if (Whitelist.TryGet(coin.Outpoint, out _))
-		{
-			return true;
-		}
+	private Dictionary<uint256, RoundVerifier> RoundVerifiers { get; } = new();
 
-		// Step 2: Check if address is from a coinjoin.
-		if (CoinJoinIdStore.Contains(coin.Outpoint.Hash))
-		{
-			return true;
-		}
-
-		return false;
-	}
-
-	public async IAsyncEnumerable<CoinVerifyInfo> VerifyCoinsAsync(IEnumerable<Coin> coinsToCheck, [EnumeratorCancellation] CancellationToken cancellationToken, string roundId = "")
+	public async IAsyncEnumerable<CoinVerifyInfo> GetCoinVerifyInfosAsync([EnumeratorCancellation] CancellationToken cancellationToken, uint256 roundId)
 	{
 		var before = DateTimeOffset.UtcNow;
 
-		using CancellationTokenSource cancellationTokenSource = new(TimeSpan.FromSeconds(30));
-		using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
-
-		var lastChangeId = Whitelist.ChangeId;
-		Whitelist.RemoveAllExpired(WabiSabiConfig.ReleaseFromWhitelistAfter);
-
-		var scriptsToCheck = new HashSet<Script>();
-		var innocentsCounter = 0;
-
-		foreach (var coin in coinsToCheck)
-		{
-			if (CheckIfAlreadyVerified(coin))
-			{
-				innocentsCounter++;
-				yield return new CoinVerifyInfo(false, coin);
-			}
-			else
-			{
-				scriptsToCheck.Add(coin.ScriptPubKey);
-			}
-		}
-
-		Logger.LogInfo($"{innocentsCounter} out of {coinsToCheck.Count()} utxo is already verified in Round({roundId}).");
-		await foreach (var response in CoinVerifierApiClient.VerifyScriptsAsync(scriptsToCheck, linkedCts.Token))
+		try
 		{
-			bool shouldBanUtxo = CheckForFlags(response.ApiResponseItem);
+			var roundVerifier = RoundVerifiers[roundId];
+			var coinVerifierTasks = roundVerifier.CloseAndGetCoinResultTasks();
 
-			// Find all coins with the same script (address reuse).
-			foreach (var coin in coinsToCheck.Where(c => c.ScriptPubKey == response.ScriptPubKey))
+			do
 			{
-				if (!shouldBanUtxo)
+				var completedTask = await Task.WhenAny(coinVerifierTasks).WaitAsync(cancellationToken).ConfigureAwait(false);
+				CoinVerifyInfo? result;
+
+				try
+				{
+					result = await completedTask.WaitAsync(cancellationToken).ConfigureAwait(false);
+				}
+				catch (CoinVerifyItemException ex)
+				{
+					result = new CoinVerifyInfo(false, true, ex.Coin);
+				}
+				catch (Exception ex)
 				{
-					Whitelist.Add(coin.Outpoint);
+					// This should never happen.
+					Logger.LogError(ex);
+					continue;
 				}
 
-				yield return new CoinVerifyInfo(shouldBanUtxo, coin);
+				yield return result;
+
+				cancellationToken.ThrowIfCancellationRequested();
 			}
+			while (coinVerifierTasks.Any(t => !t.IsCompleted));
 		}
-
-		if (Whitelist.ChangeId != lastChangeId)
+		finally
 		{
-			Whitelist.WriteToFile();
+			await Whitelist.WriteToFileIfChangedAsync().ConfigureAwait(false);
+			var duration = DateTimeOffset.UtcNow - before;
+			RequestTimeStatista.Instance.Add("verifier-period", duration);
```

Comment:
Does this line make sense? We start the requests 2 mins before the InputReg ends.

Reply:
Moved it into RoundVerifier